src/NServiceBus.Core.Tests/Recoverability/SecondLevelRetries/SecondLevelRetriesTests.cs

```diff
@@ -16,126 +16,156 @@
         }
 
         [Test]
-        public async Task ShouldRevertMessageBodyWhenDispatchingMessage()
+        public async Task ShouldNotInvokeContinuationAfterMessageFailure()
         {
-            const string originalContent = "original content";
-            var fakeDispatchPipeline = new FakeDispatchPipeline();
-            var context = CreateContext("someId", 1, fakeDispatchPipeline, Encoding.UTF8.GetBytes(originalContent));
-            var retryPolicy = new FakePolicy(TimeSpan.FromSeconds(0));
-            var behavior = new SecondLevelRetriesBehavior(retryPolicy, "test-address-for-this-pipeline");
+            var message = CreateMessage(slrRetryHeader: "1");
+            var context = CreateContext(message);
+            
+            var behavior = new SecondLevelRetriesBehavior(new FakePolicy(TimeSpan.FromSeconds(5)), string.Empty, GetFailureInfoStorage());
+            var calledTwice = false;
 
-            var message = context.Message;
-            message.Body = Encoding.UTF8.GetBytes("modified content");
 
-            await behavior.Invoke(context, () => { throw new Exception("test"); });
+            await behavior.Invoke(context, () => { throw new Exception(); });
+            await behavior.Invoke(context, () =>
+            {
+                calledTwice = true;
+                return Task.FromResult(0);
+            });
 
-            var dispatchedMessage = fakeDispatchPipeline.RoutingContext.Message;
-            Assert.AreEqual(originalContent, Encoding.UTF8.GetString(dispatchedMessage.Body));
-            Assert.AreEqual(originalContent, Encoding.UTF8.GetString(message.Body));
+            Assert.IsFalse(calledTwice, "SLR should not call pipline continuation when processing a message marked for defferal.");
         }
 
-        FakeTransportReceiveContext CreateContext(string messageId, int currentRetryCount, FakeDispatchPipeline pipeline, byte[] messageBody = null)
+        IncomingMessage CreateMessage(string id = "id", string slrRetryHeader = null, byte[] body = null)
         {
-            var context = new FakeTransportReceiveContext(messageId, currentRetryCount, messageBody);
+            var headers = string.IsNullOrEmpty(slrRetryHeader)
+                ? new Dictionary<string, string>()
+                : new Dictionary<string, string> { { Headers.Retries, slrRetryHeader }};
+
+            return new IncomingMessage(id, headers, new MemoryStream(body ?? new byte[0]));
+        }
 
-            context.Extensions.Set<IPipelineCache>(new FakePipelineCache(pipeline));
+        FakeTransportReceiveContext CreateContext()
+        {
+            return CreateContext(CreateMessage());
+        }
+
+        FakeTransportReceiveContext CreateContext(IncomingMessage message)
+        {
+            var context = new FakeTransportReceiveContext(message);
+
+            context.Extensions.Set<IPipelineCache>(new FakePipelineCache(dispatchPipline));
```

Comment:
the idea is to not the `Invoke(context, next, fork)` instead of setting up the pipeline cache